s3.1
.TH s3 1
.SH NAME
s3/cat, s3/ls, s3/rm, s3/cp, s3/write, s3/factotum
.SH SYNOPSIS
.B s3/cat
[
.B -o
.I offset
]
[
.B -n
.I count
]
.I s3://bucket/file
.br
.B s3/ls
.I s3://bucket/prefix/
.br
.B s3/rm
.I s3://bucket/file
.br
.B s3/cp
file
.I s3://bucket/destination
.br
.B s3/cp
.I s3://bucket/source
file
.br
.B s3/write
.I s3://bucket/file
.br
.B s3/write
.B -l
.br
s3/factotum
[
.B -p
]
[
.B -s
.I srvname
]
[
.B -m
.I mntpt
]
.SH DESCRIPTION
These tools provide rudimentary access to data stored within s3 (or API compatible) buckets.
The tools do little to paper over the realities of the s3 API, and mostly aim to map directly to
different HTTP verbs where applicable.
.PP
.B s3/cat
issues a GET request against the provided path and outputs the result to standard out.
.PP
.B s3/rm
issues a DELETE request against the provided path.
.PP
.B s3/ls
issues a GET request against the path, and attempts to parse the XML output as a directory listing.
.PP
.B s3/cp
copies files to and from a s3 bucket.
When conducting an upload an attempt is made to provide the correct mime type of the file through the use of
.IR file (1).
.PP
.B s3/write
reads standard input and uploads the content to a remote object chunks at a time using multipart uploads.
S3 multipart uploads are a stateful interface, they must be created, referenced, then either aborted or completed.
As such, it is feasible that one may wind up with zombie multipart uploads.
The
.B -l
flag may be used to query the list of currently open multipart uploads.
It prints out invocations of
.B s3/rm
that can be used to abort the upload.
.SH "CREDENTIALS AND PARAMETERS"
.B s3/factotum
provides an overlay factotum for handling the signing of requests using the $AWS_SECRET_ACCESS_KEY.
Each tool presents
.B s3/factotum
with the paramaters to sign, the secret access key is not directly retrievable after being stored.
.B s3/factotum
can store multiple sets of keys, and as such each secret access key must be provided with a paired
access id for disambiguation. The following is an example of adding a key pair to
.BR s3/factotum :
.PP
.EX
% echo 'key proto=aws4 access=blah !secret=whatever' >/mnt/factotum/ctl
.EE
.PP
.B s3/factotum
by default mounts itself over top of
.I /mnt/factotum
and does not post itself to
.IR /srv .
This behavior may be changed with the
.B -m
and
.B -s
flags respectfully.
.PP
Each tool needs to be provided both the access key and the endpoint, and optionally a region.
These parameters may be specified using either their standard environment variable names, or
through command line flags.
$AWS_ACCESS_KEY_ID and the
.B -k
flag may be used to specify the access id.
$AWS_ENDPOINT_URL_S3 and the
.B -u
flag may be used to specify the endpoint url.
$AWS_DEFAULT_REGION and the
.B -e
flag may be used to specify the region, defaulting to "auto" if not set.
.SH "SEE ALSO"
.IR webfs (4),
.IR factotum (4)
.SH BUGS
Only AWS4-HMAC-SHA256 signatures are supported.
Only tested against Cloudflare's R2™.
No direct s3:// → s3:// copying.


s3.c
#include <u.h>
#include <libc.h>
#include <bio.h>
#include <mp.h>
#include <libsec.h>
#include <auth.h>
#include "s3.h"

/*
 * S3 requires that all reserved characters within a query
 * are URL-encoded, despite the spec not mandating it.
 */
static int
queryvalfmt(Fmt *f)
{
	static char hex[] = "0123456789ABCDEF";
	char *s;
	int n;

	s = va_arg(f->args, char*);
	n = 0;
	while(*s != '\0'){
		switch(*s){
		case 'a'...'z':
		case 'A'...'Z':
		case '0'...'9':
			n += fmtprint(f, "%c", *s);
			break;
		default:
			n += fmtprint(f, "%%%c%c", hex[(*s)>>4], hex[(*s)&15]);
			break;
		}
		s++;
	}
	return n;
}

void
s3fmtinstall(void)
{
	fmtinstall('U', queryvalfmt);
}

typedef struct {
	uchar *payhash;
	char *mime;
	char *path;
	char *method;
	char time[128];
	char authhdr[512];
	char *range;
} Hreq;

static void
datetime(char *date, int ndate, char *time, int ntime)
{
	Tm t;

	tmnow(&t, nil);
	snprint(date, ndate, "%τ", tmfmt(&t, "YYYYMMDD"));
	snprint(time, ntime, "%sT%τZ", date, tmfmt(&t, "hhmmss"));
}

static void
getkey(char *date, char *region, char *service, char *access, uchar out[SHA2_256dlen])
{
	int fd;
	AuthRpc *rpc;
	char buf[256];
	int n;

	fd = open("/mnt/factotum/rpc", ORDWR);
	if(fd < 0)
		sysfatal("factotum rpc open: %r");
	rpc = auth_allocrpc(fd);
	n = snprint(buf, sizeof buf, "proto=aws4 access=%s", access);
	if(auth_rpc(rpc, "start", buf, n) != ARok)
		sysfatal("auth_rpc: %r");
	n = snprint(buf, sizeof buf, "%s %s %s", date, region, service);
	if(auth_rpc(rpc, "write", buf, n) != ARok)
		sysfatal("auth_rpc: %r");
	if(auth_rpc(rpc, "read", nil, 0) != ARok)
		sysfatal("auth_rpc: %r");
	if(rpc->narg != SHA2_256dlen)
		sysfatal("invalid auth_rpc output");
	memcpy(out, rpc->arg, SHA2_256dlen);
	auth_freerpc(rpc);
	close(fd);
}

static void
mkreq(Hreq *hreq, char *method, char *path, uchar *payhash, char *mime)
{
	hreq->method = method;
	hreq->path = path;
	hreq->payhash = payhash;
	hreq->mime = mime;
	hreq->range = nil;
}

static void
signreq(Hreq *hreq, S3 *s3)
{
	char date[64];
	uchar key[SHA2_256dlen], sig[SHA2_256dlen];
	char buf[1024], buf2[1024], req[1024];
	char *sgndhdr;
	char *query;

	datetime(date, sizeof date, hreq->time, sizeof hreq->time);
	if(strcmp(hreq->method, "POST") == 0){
		snprint(buf, sizeof buf, "content-type:%s\nhost:%s\nx-amz-content-sha256:%.*lH\nx-amz-date:%s\n",
			hreq->mime, s3->host, SHA2_256dlen, hreq->payhash, hreq->time);
		sgndhdr = "content-type;host;x-amz-content-sha256;x-amz-date";
	} else if(strcmp(hreq->method, "PUT") == 0){
		snprint(buf, sizeof buf, "content-type:%s\nhost:%s\nx-amz-content-sha256:%.*lH\nx-amz-date:%s\n",
			hreq->mime, s3->host, SHA2_256dlen, hreq->payhash, hreq->time);
		sgndhdr = "content-type;host;x-amz-content-sha256;x-amz-date";
	} else if(strcmp(hreq->method, "GET") == 0 || strcmp(hreq->method, "DELETE")==0 || strcmp(hreq->method, "POST") == 0){
		sha2_256(nil, 0, hreq->payhash, nil);
		snprint(buf, sizeof buf, "host:%s\nx-amz-date:%s\n", s3->host, hreq->time);
		sgndhdr = "host;x-amz-date";
	} else
		sysfatal("invalid method");

	snprint(buf2, sizeof buf2, "%s", hreq->path);
	if((query = strchr(buf2, '?')) != nil){
		*query = '\0';
		query++;
	} else
		query = "";
	snprint(req, sizeof req, "%s\n/%s/%s\n%s\n%s\n%s\n%.*lH",
		hreq->method, s3->bucket, buf2, query, buf, sgndhdr, SHA2_256dlen, hreq->payhash);
	sha2_256((uchar*)req, strlen(req), key, nil);
	snprint(buf, sizeof buf, "%s\n%s\n%s/%s/%s/aws4_request\n%.*lH",
		"AWS4-HMAC-SHA256", hreq->time, date, s3->region, "s3", SHA2_256dlen, key);
	getkey(date, s3->region, "s3", s3->access, key);
	hmac_sha2_256((uchar*)buf, strlen(buf), key, SHA2_256dlen, sig, nil);

	snprint(hreq->authhdr, sizeof hreq->authhdr, "%s Credential=%s/%s/%s/%s/aws4_request, SignedHeaders=%s, Signature=%.*lH",
		"AWS4-HMAC-SHA256", s3->access, date, s3->region, "s3", sgndhdr, SHA2_256dlen, sig);
}

/* small fprint buffers bite us */
#pragma	   varargck    argpos	   ctlprint 2
static long
ctlprint(int cfd, char *fmt, ...)
{
	char buf[2048];
	char *e;
	va_list arg;

	va_start(arg, fmt);
	e = vseprint(buf, buf + sizeof buf, fmt, arg);
	va_end(arg);
	return write(cfd, buf, e-buf);
}

static int
prep(S3 *s3, int cfd, Hreq *hreq)
{
	if(ctlprint(cfd, "url %s/%s/%s", s3->endpoint, s3->bucket, hreq->path) < 0)
		return -1;
	if(ctlprint(cfd, "request %s", hreq->method) < 0)
		return -1;
	if(ctlprint(cfd, "headers Authorization:%s", hreq->authhdr) < 0)
		return -1;
	if(ctlprint(cfd, "headers x-amz-date:%s\nx-amz-content-sha256:%.*lH", hreq->time, SHA2_256dlen, hreq->payhash) < 0)
		return -1;
	if(hreq->mime != nil && ctlprint(cfd, "contenttype %s", hreq->mime) < 0)
		return -1;
	if(hreq->range != nil && ctlprint(cfd, "headers range: %s", hreq->range) < 0)
		return -1;
	return 0;
}

static int
hopen(Hcon *h, S3 *s3, int mode, Hreq *hreq)
{
	long n;
	char buf[64];

	h->body = -1;
	h->post = -1;
	h->err = -1;
	h->ctl = open("/mnt/web/clone", ORDWR);
	if(h->ctl < 0)
		return -1;

	n = read(h->ctl, h->id, sizeof h->id - 1);
	if(n <= 0){
		close(h->ctl);
		werrstr("short read from /mnt/web/clone");
		return -1;
	}
	h->id[n-1] = 0;

	signreq(hreq, s3);
	if(prep(s3, h->ctl, hreq) < 0){
		close(h->ctl);
		return -1;
	}
	switch(h->mode = mode){
	case OREAD:
		snprint(buf, sizeof buf, "/mnt/web/%s/body", h->id);
		h->body = open(buf, OREAD);
		if(h->body >= 0)
			return 0;
		snprint(buf, sizeof buf, "/mnt/web/%s/errorbody", h->id);
		h->err = open(buf, OREAD);
		return -1;
	case ORDWR:
	case OWRITE:
		snprint(buf, sizeof buf, "/mnt/web/%s/postbody", h->id);
		h->post = open(buf, OWRITE);
		return 0;
	default:
		return -1;
	}
}

void
hclose(Hcon *h)
{
	if(h->ctl >= 0)
		close(h->ctl);
	if(h->body >= 0)
		close(h->body);
	if(h->post >= 0)
		close(h->post);
	if(h->err >= 0)
		close(h->err);
}

int
hdone(Hcon *h)
{
	char buf[64];

	switch(h->mode){
	case OWRITE:
	case ORDWR:
		close(h->post);
		h->post = -1;
		snprint(buf, sizeof buf, "/mnt/web/%s/body", h->id);
		h->body = open(buf, OREAD);
		if(h->body < 0){
			snprint(buf, sizeof buf, "/mnt/web/%s/errorbody", h->id);
			h->err = open(buf, OREAD);
			return -1;
		}
		return 0;
	default:
		abort();
	}
}

int
s3get(S3 *s3, Hcon *con, char *path)
{
	Hreq h;
	uchar payhash[SHA2_256dlen];

	mkreq(&h, "GET", path, payhash, nil);
	return hopen(con, s3, OREAD, &h);
}

int
s3getrange(S3 *s3, Hcon *con, char *path, long off, long n)
{
	Hreq h;
	uchar payhash[SHA2_256dlen];
	char buf[64];

	mkreq(&h, "GET", path, payhash, nil);
	snprint(buf, sizeof buf, "bytes=%ld-%ld", off, off+n-1);
	h.range = buf;
	return hopen(con, s3, OREAD, &h);
}

int
s3put(S3 *s3, Hcon *con, char *path, char *mime, uchar *payhash)
{
	Hreq h;

	mkreq(&h, "PUT", path, payhash, mime);
	return hopen(con, s3, ORDWR, &h);
}

int
s3del(S3 *s3, Hcon *con, char *path)
{
	Hreq h;
	uchar payhash[SHA2_256dlen];

	mkreq(&h, "DELETE", path, payhash, nil);
	return hopen(con, s3, OREAD, &h);
}

int
s3post(S3 *s3, Hcon *con, char *path)
{
	Hreq h;
	uchar payhash[SHA2_256dlen];

	sha2_256(nil, 0, payhash, nil);
	mkreq(&h, "POST", path, payhash, "application/octet-stream");
	if(hopen(con, s3, ORDWR, &h) < 0)
		return -1;
	
	return hdone(con);
}

int
s3postwrite(S3 *s3, Hcon *con, char *path, char *mime, uchar *payhash)
{
	Hreq h;

	mkreq(&h, "POST", path, payhash, mime);
	return hopen(con, s3, ORDWR, &h);
}
